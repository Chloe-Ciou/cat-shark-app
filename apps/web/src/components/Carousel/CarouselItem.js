import PropTypes from 'prop-types';

const CarouselItem = ({ slide, isActive, index }) => {
  return (
    <div data-testid={`slide_${index}`} className={`slide fade ${isActive ? 'active' : ''}`}>
      <img src={slide.src} alt={slide.alt} className="image" />
    </div>
  )
}

CarouselItem.propTypes = {
  index: PropTypes.number,
  isActive: PropTypes.bool,
  slide: PropTypes.shape({
    src: PropTypes.string,
    alt: PropTypes.string,
  })
};

export default CarouselItem;