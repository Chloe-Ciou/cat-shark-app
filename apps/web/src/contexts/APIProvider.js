import { createContext, useContext } from "react";
import PropTypes from 'prop-types';

import apiClient from "../utils/apiClient";

export const APIContext = createContext();

const APIProvider = ({ children }) => {
  return <APIContext.Provider value={apiClient}>{children}</APIContext.Provider>
};

APIProvider.displayName = 'APIProvider';
APIProvider.propTypes = {
  children: PropTypes.node,
};

export default APIProvider;

/**
 * Connects a component to the <APIProvider />
 * @returns {APIProvider}
 */
export function useAPI() {
  const context = useContext(APIContext);

  if (typeof context === 'undefined') {
    throw new Error('useAPI must be used within a APIProvider');
  }

  return context;
}