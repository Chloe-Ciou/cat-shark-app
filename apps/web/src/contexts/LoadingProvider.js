import { useState, createContext, useContext } from "react";
import PropTypes from 'prop-types';

export const LoadingContext = createContext();

const LoadingProvider = ({ children }) => {
  const [isLoading, setIsLoading] = useState(true);

  return <LoadingContext.Provider value={{ isLoading, setIsLoading }}>{children}</LoadingContext.Provider>
};

LoadingProvider.displayName = 'LoadingProvider';
LoadingProvider.propTypes = {
  children: PropTypes.node,
};

export default LoadingProvider;

/**
 * Connects a component to the <LoadingProvider />
 * @returns {LoadingProvider}
 */
export function useLoading() {
  const context = useContext(LoadingContext);

  if (typeof context === 'undefined') {
    throw new Error('useLoading must be used within a LoadingProvider');
  }

  return context;
}