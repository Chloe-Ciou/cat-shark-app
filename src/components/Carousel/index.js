import { useState, useEffect } from "react";

import CarouselItem from "./CarouselItem";
import { useImages } from "../../hooks/useImages";
import { IMAGE_CATEGORIES } from "../../constants";

const Carousel = () => {
  const { getImages, images } = useImages();
  // In order to handle more image types in the future, thus array type is chosen here.
  const [imgTypes, setImgTypes] = useState([IMAGE_CATEGORIES.CAT.type]);
  const [currentSlideIndex, setCurrentSlideIndex] = useState(0);
  const slidesLength = images?.length || 0;

  useEffect(() => {
    if (imgTypes.length) {
      getImages(imgTypes, () => setCurrentSlideIndex(0));
    }
  }, [imgTypes, getImages, setCurrentSlideIndex]);

  const checkIsButtonActive = (type) => imgTypes.find((imgType) => type === imgType);
  const renderPrevSlide = () => setCurrentSlideIndex(currentSlideIndex ? currentSlideIndex - 1 : slidesLength - 1);
  const renderNextSlide = () => setCurrentSlideIndex(currentSlideIndex === slidesLength - 1 ? 0 : currentSlideIndex + 1);

  /**
   * Sets image type list.
   * @param {String} imageType 'cat' | 'shark'
   */
  const setImageTypes = (imageType) => {
    const index = imgTypes.findIndex((type) => type === imageType);
    if (index === -1) {
      setImgTypes([...imgTypes, imageType]);
    } else if (index !== -1 && imgTypes.length > 1) {
      imgTypes.splice(index, 1);
      setImgTypes([...imgTypes]);
    }
  };

  const renderButtonGroup = () => {
    return Object.values(IMAGE_CATEGORIES).map((category) =>
      <button data-testid={`category_btn_${category.type}`} key={category.type} className={`${checkIsButtonActive(category.type) ? "active" : ""}`} onClick={() => setImageTypes(category.type)}>{category.name}</button>);
  };

  return (
    <section className="carousel" data-testid="carousel">
      <div className="slides">
        <div className="button-group">{renderButtonGroup()}</div>
        {slidesLength
          ? images.map((slide, index) => <CarouselItem key={`${slide.alt}_${index}`} index={index} slide={slide} isActive={currentSlideIndex === index} />)
          : <div className="no-content">No photos</div>
        }
        <span className="prev">
          <button data-testid="prev_btn" aria-label="previous slide" onClick={renderPrevSlide}>&#10094;</button>
        </span>
        <span className="next">
          <button data-testid="next_btn" aria-label="next slide" onClick={renderNextSlide}>&#10095;</button>
        </span>
      </div>
    </section >
  );
};

export default Carousel;
