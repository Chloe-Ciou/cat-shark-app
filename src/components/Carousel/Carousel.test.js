import { render, screen, waitFor, fireEvent } from "@testing-library/react";

import Carousel from "./index";
import APIProvider from "../../contexts/APIProvider";
import LoadingProvider from "../../contexts/LoadingProvider";

jest.mock("../../hooks/useImages", () => ({
  useImages: () => ({
    getImages: jest.fn(),
    images: [{ src: "src_1", alt: "alt_1" }, { src: "src_2", alt: "alt_2" }, { src: "src_3", alt: "alt_3" }]
  })
}));

describe("<Carousel />", () => {
  it("should match snapshots", () => {
    renderCarousel();
    expect(screen.getByTestId("carousel")).toMatchSnapshot();
  });

  it("should change active slide when prev/next btn was clicked", async () => {
    renderCarousel();
    expect(screen.getByTestId("slide_0")).toHaveClass("active");
    expect(screen.getByTestId("slide_1")).not.toHaveClass("active");
    fireEvent.click(screen.getByTestId("next_btn"));

    await waitFor(() => expect(screen.getByTestId("slide_0")).not.toHaveClass("active"));
    await waitFor(() => expect(screen.getByTestId("slide_1")).toHaveClass("active"));

    fireEvent.click(screen.getByTestId("prev_btn"));
    await waitFor(() => expect(screen.getByTestId("slide_0")).toHaveClass("active"));
    await waitFor(() => expect(screen.getByTestId("slide_1")).not.toHaveClass("active"));
  });

  it("should change button active state when btn was clicked", async () => {
    renderCarousel();
    expect(screen.getByTestId("category_btn_cat")).toHaveClass("active");
    expect(screen.getByTestId("category_btn_shark")).not.toHaveClass("active");

    fireEvent.click(screen.getByTestId("category_btn_shark"));
    await waitFor(() => expect(screen.getByTestId("category_btn_shark")).toHaveClass("active"));

    fireEvent.click(screen.getByTestId("category_btn_cat"));
    await waitFor(() => expect(screen.getByTestId("category_btn_cat")).not.toHaveClass("active"));
    await waitFor(() => expect(screen.getByTestId("category_btn_shark")).toHaveClass("active"));
  });
});

const renderCarousel = () => {
  return render(
    <LoadingProvider>
      <APIProvider>
        <Carousel />
      </APIProvider>
    </LoadingProvider>
  );
};
