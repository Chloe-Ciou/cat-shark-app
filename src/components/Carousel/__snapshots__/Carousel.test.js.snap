// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> should match snapshots 1`] = `
<section
  class="carousel"
  data-testid="carousel"
>
  <div
    class="slides"
  >
    <div
      class="button-group"
    >
      <button
        class="active"
        data-testid="category_btn_cat"
      >
        Cats
      </button>
      <button
        class=""
        data-testid="category_btn_shark"
      >
        Sharks
      </button>
    </div>
    <div
      class="slide fade active"
      data-testid="slide_0"
    >
      <img
        alt="alt_1"
        class="image"
        src="src_1"
      />
    </div>
    <div
      class="slide fade "
      data-testid="slide_1"
    >
      <img
        alt="alt_2"
        class="image"
        src="src_2"
      />
    </div>
    <div
      class="slide fade "
      data-testid="slide_2"
    >
      <img
        alt="alt_3"
        class="image"
        src="src_3"
      />
    </div>
    <span
      class="prev"
    >
      <button
        aria-label="previous slide"
        data-testid="prev_btn"
      >
        ❮
      </button>
    </span>
    <span
      class="next"
    >
      <button
        aria-label="next slide"
        data-testid="next_btn"
      >
        ❯
      </button>
    </span>
  </div>
</section>
`;
